Remove duplicated code in sound/soc/bcm soundcard drivers.

timg236 opened the pull request:
There's quite a lot of duplicated code in the Raspberry Pi soundcard drivers and since they aren't upstreamed it's a bit of a maintenance burden. 

This series of commits firstly does some general refactoring to make the code less verbose by using devm for card registration and moves static bclk configuration to init instead of implementing a hw_params callback.

The second and larger change introduces a generic driver for simple cards. The simple card drivers don't really do much more than set bclk but unfortunately there's no way to do that via the soc/generic cards preventing them from becoming "DT only" drivers.

This version adds support for ADAU1977 ADC, Google VoiceHat, HifiBerry AMP, HifiBerry DAC and RPI DAC.

TODO
There's a middle ground for drivers which are slightly more complex. It might be possible to create utils functions for common features e.g. muting via gpio, volume limiting. However, I think it's better to get the simple fixes in first.

Commit:
ASoC: bcm: Refactor to use devm to register cards

The .remove function in many of the BCM drivers just called
snd_soc_unregister_card. Migrate these drivers to use devm to manage
card registration allowing the .remove functions to be eliminated.

Drivers performing other operations in .remove have not been updated.

Signed-off-by: Tim Gover <tim.gover@raspberrypi.org>

Commit:
ASoC: rpi-dac: Move snd_soc_dai_set_bclk_ratio to .init

snd_soc_dai_set_bclk_ratio is called with a fixed value of 64. This can
be called from .init instead allowing the more verbose hw_params
callback method to be removed.

Signed-off-by: Tim Gover <tim.gover@raspberrypi.org>

Commit:
ASoC: hifiberry_api: Move snd_soc_dai_set_bclk_ratio to .init

snd_soc_dai_set_bclk_ratio is called with a fixed value of 64.
This can be called from .init instead allowing the more verbose
hw_params callback method to be removed.

Signed-off-by: Tim Gover <tim.gover@raspberrypi.org>

Commit:
ASoC: Add generic RPI driver for simple soundcards.

The RPI simple sound card driver provides a generic ALSA SOC card driver
supporting a variety of Pi HAT soundcards. The intention is to avoid
the duplication of code for cards that can't be fully supported by
the soc simple/graph cards but are otherwise almost identical.

This initial commit adds support for the ADAU1977 ADC, Google VoiceHat,
HifiBerry AMP, HifiBerry DAC and RPI DAC.

Signed-off-by: Tim Gover <tim.gover@raspberrypi.org>

popcornmix commented:
Ping @HiassofT - interested in your comments.

HiassofT commented on sound/soc/bcm/Kconfig:
```diff
@@ -196,3 +196,16 @@ config SND_PISOUND
+	select SND_SOC_ADAU1977_I2C
+	select SND_SOC_PCM1794A
+	select SND_SOC_PCM5102A
+	select SND_SOC_PCM512x_I2C
```
PCM512x doesn't seem to be used by the drivers and can be dropped

timg236 replied:
Will do. That's some old config from when I was looking at the iqaudio-dac which isn't covered by RPi simple

HiassofT commented:
thanks, looks fine, only thing that stuck out was the PCM512x_I2S selection in Kconfig which can be dropped (unless I missed something)

pelwell commented:
I was wondering if we could keep the old per-soundcard config settings but make them select the rpi-simple driver (which then wouldn't need a menu entry). The Kconfig entry for each card would select its own codec and our driver, and the rpi-simple driver could have codec support predicated on the soundcard config settings.

I think this scheme would allow our umbrella driver to support a single soundcard without adding any significant bloat over the dedicated driver, and the full-fat Raspbian build would be identical to this PR.

HiassofT commented:
Setting up Kconfig that way is a good idea. If we keep settings as is (and only make them select the generic drivers) we also get a seamless upgrade path via make oldconfig

Commit:
ASoC: Enable the RPi simple soundcard driver

Switch the following soundcards to use the RPi simple soundcard driver
and remove the dedicate driver implementations.

* ADAU1977 ADC
* Google VoiceHat
* HifiBerry AMP
* HifiBerry DAC
* RPI DAC

There is no impact on defconfig or codec configuration. Instead of
enabling the config to build the dedicated driver the menuconfig entries
now just select SND_RPI_SIMPLE_SOUNDCARD.

Signed-off-by: Tim Gover <tim.gover@raspberrypi.org>

timg236 commented on sound/soc/bcm/Makefile:
```diff
@@ -1,4 +1,4 @@
-# BCM2835 Platform Support
+# BCM2836 Platform Support
```
Remove vim ctrl+a typo :)

Commit:
ASoC: Create a generic Pi Hat WM8804 driver

Reduce the amount of duplicated code by creating a generic driver for
Pi Hat digi cards using the WM8804 codec.

This replaces the
Allo DigiOne, Hifiberry Digi/Pro, JustBoom Digi and IQAudIO Digi
dedicate soundcard drivers with a generic driver.

There are no significant changes to the runtime behavior of the drivers
and end users should not have to change any configuration settings
after upgrading.

Minor changes
* Check the return value of snd_soc_component_update_bits
* Added some pr_debug tracing
* Various checkpatch tidyups
* Updated allodigi-one to use use 128FS at > 96 Khz. This appears to
  be an omission in the original driver code so followed the Hifiberry
  DAC driver approach.

Commit:
ASoC: Use correct card name in rpi-simple driver

Use the specific card name from drvdata instead of the snd_rpi_simple

timg236 commented:
I've added an extra commit which adds a generic WM8804 driver replacing JustBoom Digi, IQAudIO Digi, HifiBerry Digi and Allo DigiOne cards.

As with the rpi-simple driver this should not be visible to end users or require any defconfig changes.

Looking for feedback from the driver authors before posting  this on the forums.

I have tested with the JustBoom digi soundcard. Unfortunately, I don't have the other cards to test with at the moment.

pelwell commented:
@Hifiberry @iqaudio @shawaj @allocom @PeterMalkin Any feedback?

hifiberry commented:
This won't work as the whole clock-selection code has been removed.
 Both the DAC+ Pro and the Digi+ Pro are not "generic" cards, but use their own clock generators that are controlled by the existing drivers. 

While I like the idea to have as much code as possible moved to a generic driver, there is a need for adoptions for a specific hardware.

hifiberry commented:
@timg236 : Please contact me at info@hifiberry.com if you need samples for development.

timg236 commented:
@hifiberry the DAC+Pro driver should not be changed by these patches. 

The HifiBerry Digi+ and Allo Digione should both have the correct clock selection from snd_rpi_wm8804_probe unless I've missed something.

hifiberry commented:
@timg236 : Right, the DAC+ driver isn't affected, but the Digi+ is. There is hardware-specific clock control code (check lines 43 and later in the Digi+ driver) that is not part of a generic WM8804 driver. As far as I see this code is simply remove by this patch.

timg236 commented:
@hifiberry snd_rpi_hifiberry_digi_enable_clock and moved to snd_rpi_wm8804_enable_clock at line 112.

It should be following the same logic of checking for "clock44" and "clock48" from devm and using that to distinguish between the Digi and the Digi+. The Allo DigiOne seemed to have almost identical logic so I basically renamed the functions to be generic.

Although on the Allo DigiOne there is only one variant so the absence of clock44 and clock48 are an error.

For the IQAudIO and JustBoom Digi there's no special clock configuration so the code just ignores the absence of clock44/clock48.

The design was to use the snd_rpi_wm8804_drvdata to encapsulate configuration data or optional callbacks for hardware specific customisation so there's a fair amount of wiggle room there if we need to change things.

hifiberry commented:
@timg236 I can't find this. What file did you include this function (snd_rpi_wm8804_enable_clock) ?

hifiberry commented:
Ok, now I found the file. It was hidden in the Github UI :(
I'll make sure that you get sample boards soon.

allocom commented:
Hi Tim,  I will test and update the result with our Allo digione.
For further, development & testing on our side, you can write to Andre (Andre@allo.com) for a sample.

HiassofT commented:
A quick test with a Hifiberry Digi Pro looks fine - 44.1, 48, 88.2 and 96kHz work nicely. So clock switching seems to be OK

Tested on RPi3 with current rpi-4.18.y tree plus this PR, using bcm2709_defconfig

allocom commented:
On quick testing with Allo Digione, it is working fine.

timg236 commented:
Rebased - no changes to the patches

pelwell commented:
Thanks for the feedback. As 4.18 is hardly mainstream yet I'm happy to merge and let a few early adopters kick the tyres.

PeterMalkin commented:
Thank you! We have tested on VoiceHAT - works great!
